### ISSUE TITLE
Data loss and other save issues w/0.9.41 and 0.9.42

### ISSUE THREAD

**Issue by pjeby** [typora/typora-issues#1184]
I recently upgraded from 0.9.37 to 0.9.41 and started having issues with other tools, and problems losing files.  Today I lost an entire day's work because Typora stopped being able to save a file and it looked as though it saved but in fact it *deleted* the original file.

There seems to be something else weird about the saving process as well:  I normally run the [`entr` file monitoring tool](http://entrproject.org/) on my server to detect changes to the files I'm editing with Typora, in order to re-run the tests that are embedded in the files.  Upon upgrading to 0.9.41 and above this stopped working: entr no longer detects when the files are saved.

It also used to be the case that opening a file in another editor besides Typora would "play nicely", in that Typora would pick up the changes.  Now, it seems to be holding something open or locking, because saving in another editor gives permission errors on the file.

Between these various problems I've had to revert to 0.9.37 because entr is a critical part of my development workflow.  (I tried upgrading to 0.9.42, and it has the same problem.)

Running inotifywait to investigate, I notice that while 0.9.37 appears to close the file after saving, 0.9.41 and 0.9.42 appear to reopen it and keep it open.  (This is probably what's keeping entr from working properly, as it expects to see the file closed for a bit before it considers it safe to run the build and test process on it.)

The other problem with keeping the file open is that I'm using a network connection over a VPN that sometimes lags or drops, and I suspect that may have been involved in the data loss issue, even though the connection was working at the time I attempted to save.  There is sometimes an issue where Typora will appear to have saved the file, but inspecting the file on the other end of the connection it remains unchanged.  I don't know if that's the same issue that caused the file to disappear altogether today, though.  (As in, I told Typora to save the file, and then the file ceased to exist, forcing me to restore a day-old backup.)

**Comment by abnerlee** [comment]
What's the OS?

Will save fail when on other apps are editing or monitoring the file?

**Comment by pjeby** [comment]
I'm on Windows 7 64 bit, operating over a Samba connection over a VPN to a Linux server.  The other editor I tried was jEdit, also running on Windows over the same Samba+VPN connection.

(Thanks for the tip re: version control: I was able to find an up-to-date copy of the file I lost!)

I sent you a log in email.  I did previously confirm that both 0.9.41 and 0.9.42 were behaving the same as far as what inotifywait was showing after the first save.  Both of them would show a series of writes and closes followed by ATTRIB and then ACCESS, and then no subsequent close until the next save.

What's really weird, though, is that I just upgraded to 0.9.42 a second time and it's *not* happening now: entr is picking up the changes.  But...  the *first* time the file was opened, I got this access pattern:

```
# inotifywait -m *.md
Setting up watches.
Watches established.
docker-compose.jq.md OPEN
docker-compose.jq.md CLOSE_NOWRITE,CLOSE
docker-compose.jq.md CLOSE_NOWRITE,CLOSE
docker-compose.jq.md OPEN
docker-compose.jq.md CLOSE_NOWRITE,CLOSE
docker-compose.jq.md OPEN
docker-compose.jq.md ACCESS
docker-compose.jq.md MODIFY
```

IOW, the file was left open at that point as far as I can tell.  After switching between windows and saving, however, the file was then properly closed after saves.

But...  once I opened the file in jEdit, then saved it, I then went back to Typora and tried to save, and temporarily got the problem where it gives no visible indication of error, but does nothing to the actual file on the server.  But then something I did made it go away and I haven't been able to reproduce it again.

At this point I have not been able to reproduce my .41 issues with .42, except for that momentary issue, which I haven't been able to re-reproduce.  So I'm not terribly sure how to proceed.   I guess for the moment I'll stick with .42 and see if I can figure out a pattern.

**Comment by abnerlee** [comment]
Since v0.9.41 is already out-of-date, please only confirm on v0.9.42

>  once I opened the file in jEdit, then saved it, I then went back to Typora and tried to save, and temporarily got the problem where it gives no visible indication of error, but does nothing to the actual file on the server

So this is on v0.9.42, right? 
And when you go back to Typora, did Typora reload the content automatically (since the content is already modified by jEdit).
And is this relates to "the first time the file was opened"?

And about file content not updated after save, is it possible that due to the network delay, when you check it the second time, actually the save process has not been finished?

**Comment by abnerlee** [comment]
relates https://github.com/typora/typora-issues/issues/1058

**Comment by pjeby** [comment]
The only thing that I have seen more than once on .42 is entr not recognizing file changes, but I have yet to figure out what actions are required to cause that.  It happened once when I first updated to that version, and once after I updated the second time.  I have not been able to reproduce it again, though I suppose I could uninstall and reinstall it again.  I will just have to observe things going forward and if it happens again I will try to see what I can do to reproduce it intentionally.
